### ISSUE TITLE
Compose not firing when you open an in-progress reply in a new window

### ISSUE THREAD

**Issue by KennethScott** [KartikTalwar/gmail.js#392]
I've written a small Chrome extension that adds a button to the compose window's toolbar.  Everything works as expected, except in one odd scenario where the user is replying to an email and they decide to click the "In new window" icon to pop it out to a new window:

![image](https://cloud.githubusercontent.com/assets/823499/25203200/f8d2604e-251d-11e7-84b6-f1f7a194610b.png)

If the user has already clicked Reply and started their Compose _and has already entered text_ into the compose section when they click the new window icon - then the window loads and the compose section appears with their text - but the compose event **never** fires (and my button is not added).  The key is that the compose section must exist **and** there must be text in the response when they click "In new window" - otherwise when the new window opens, compose won't actually have been started and it's not an issue.  They'd just click Reply and start a new Compose section and everything works as expected.

I can see that loaded obviously fires as expected when this new window is loaded, so I'm not sure why compose doesn't.  It's like it knows it already existed back on the main page, and doesn't bother firing again on the new window (even though it seems like it needs to).  

There's not much to my code, basically just listening for compose in order to add the button:

````javascript
	  gmail.observe.on('compose', function(compose_ref) {  	  
		gmail.tools.add_compose_button(compose_ref, 'My Button', function() {
		  // do button stuff...		  
		}, 'myButtonClass'); 
	  });  
````

I've also tried wrapping the above code in an on.load to ensure everything is ready, but it didn't help:
````gmail.observe.on('load', function(){...});````

So TL;DR to reproduce-  Find an email in your Inbox, click reply, type some text, click the "In new window" icon.  When the new window appears, the compose event will not have fired (even though the compose section will be visible with your text).

Any ideas?

**Comment by josteink** [comment]
I can confirm that *something* is going on here, although I can't reproduce your exact scenario.

In my scenariom, I add some extra buttons in the email-menu (per email) when the email is shown, and I can't see *this* getting run in child-windows (although I have some suspicions on the exact reasons for this).

But for me, the `compose` observer is run every time, in all Windows. Could it be how you load your extension into the Gmail DOM?

I would advice either adding more logging to see what runs (and thus where it stops running, i.e. fails) or add debugging breakpoints and stepping your way through to see if anything fishy goes on.

**Comment by KennethScott** [comment]
I stripped down the code to a very basic example.  Maybe it does have something to do with how I'm loading it into the DOM - I'm going by the [boilerplate demo](https://github.com/KartikTalwar/gmail-chrome-extension-boilerplate).  Maybe I'm doing something stupid.

Here's the stripped down code in its entirety with some additional console.log()'s:

````javascript
var gmail;

function refresh(f) {
    console.log('refreshing');

    // boilerplate check doesn't work when child window loaded.. get Gmail is not defined error thrown.
    //if ((/in/.test(document.readyState)) || (typeof Gmail === undefined)) {
    if ((/in/.test(document.readyState)) || (typeof Gmail == 'undefined')) {
        console.log('setting timeout');
        setTimeout('refresh(' + f + ')', 10);
    } else {
        console.log('calling main');
        f();
    }
}

var main = function() {
    console.log('running main');

    gmail = new Gmail();

    gmail.observe.on('load', function() {
        console.log('loaded');

        gmail.observe.on('compose', function(compose_ref) {
            console.log("composed... now adding button...");

            gmail.tools.add_compose_button(compose_ref, 'My Button', function() {
                // do button click stuff...
            }, 'myButtonClass'); // css class for button added

        });
    });
}

refresh(main);
````

When the GMail Inbox fully loads, everything looks good.  This is the tail of the console.log()'s:
![image](https://cloud.githubusercontent.com/assets/823499/25260640/49b6bde8-2613-11e7-9839-734b9f4f2518.png)

Now I select an email, click Reply to start the compose, and see compose fire in the console:
![image](https://cloud.githubusercontent.com/assets/823499/25260682/86a9b9f8-2613-11e7-86e4-45cebfa7e0e9.png)

So everything is good - I see my button was added and add some reply text:
![image](https://cloud.githubusercontent.com/assets/823499/25260757/e1f3bf34-2613-11e7-9f84-2d1adf55856f.png)

Now I click the **In new window** icon to pop it out to a child window, and I see my button is now missing: 
![image](https://cloud.githubusercontent.com/assets/823499/25260834/57289842-2614-11e7-945f-7c56e6e0ccac.png)

Here's the console.log()'s from the child window:
![image](https://cloud.githubusercontent.com/assets/823499/25260870/87826630-2614-11e7-8a39-4d3d40ebba36.png)

Which look normal up through loaded, but compose never fires for some reason.

Am I just doing something wrong?

Do note that I deviated from the boilerplate code in how it checks for Gmail === undefined.  That just wouldn't work for me in a child window.  I had to fall back to the old school typeof Gmail == 'undefined' for it to realize it wasn't loaded yet and keep looping.  Could that have something to do with it?

It's probably worth noting that from the child window, if I discard my in-progress reply and then click Reply again (to start a new compose) - compose does fire that time and my button is added.

**Comment by josteink** [comment]
Thanks for a very detailed reply. Based on your response, it definitely seems like a load-order thing to me. 

I would suggest try adding the compose observer at the same time as the load observer. 

If that doesn't work, try again using the node-boilerplate example instead. It should further alleviate loading-related issues.

**Comment by KennethScott** [comment]
OK, my intent was to wait on load before adding anything else.  That's why I was nesting them:
````javascript
    gmail.observe.on('load', function() {
        gmail.observe.on('compose', function(compose_ref) {....});
    });
````
But you're saying just add the compose observer at the same time like this:
````javascript
    gmail.observe.on('load', function() {});
    gmail.observe.on('compose', function(compose_ref) {....});
````
I guess I don't even need load then.  I'll play around with it and see what happens.

I also looked at your node-boilerplate, but I had a hard time comparing apples to apples considering this is for a Chrome extension vs the node setup.  The gist I got from it was you are programmatically injecting your script tags, which looks like what the chrome-extension-boilerplate does in content.js:

````javascript
var j = document.createElement('script');
j.src = chrome.extension.getURL('jquery-1.10.2.min.js');
(document.head || document.documentElement).appendChild(j);

var g = document.createElement('script');
g.src = chrome.extension.getURL('gmail.js');
(document.head || document.documentElement).appendChild(g);

var s = document.createElement('script');
s.src = chrome.extension.getURL('main.js');
(document.head || document.documentElement).appendChild(s);
````

The only other piece is the manifest.json setting up the content_script to inject the web_accessible_resources:
````json
  "content_scripts": [
    {
      "matches": ["https://mail.google.com/*"],
      "js": ["content.js"],
      "css" : ["main.css"]
    }
  ],
  "web_accessible_resources": [
    "jquery-1.10.2.min.js",
    "gmail.js",
    "main.js"
  ],
````
So I'm still a little fuzzy about how I can better change the load order.  Do you see any obvious issues with what I'm doing?  Thanks again-

**Comment by josteink** [comment]
I think you misunderstand. The Node-example is for building an extension too. 

The difference is that it uses NodeJS-style require() statements to handle dependencies and does script bundling, which results in one JS you inject into the page which contains everything you need, instantly loaded. 

It's much easier to reason about than trying to load three scripts in random order. 

I suggest you check out the example a little closer.

**Comment by KennethScott** [comment]
Sorry - yes, I completely missed the boat on what you were trying to tell me.   

I switched over to your node-based example extension and everything is working perfectly in the child windows now!

Thank you for your patience!

**Comment by josteink** [comment]
Glad to hear you managed to solve your issue. Feel free to report any other issues you encounter in the future!
